File: 865-smallest-subtree-with-all-the-deepest-nodes/865-smallest-subtree-with-all-the-deepest-nodes.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int findDepth(TreeNode* node) {
        
        if (!node->left && !node->right) return 0;
        
        int ld = node->left ? findDepth(node->left) : 0;
        int rd = node->right ? findDepth(node->right) : 0;
        return 1 + max(ld, rd);
    }
    
    TreeNode* getSubtree(TreeNode* node, int curDepth, const int& targetDepth) {
        
        if (curDepth == targetDepth || !node) return node;
        
        TreeNode* left = getSubtree(node->left, curDepth + 1, targetDepth);
        TreeNode* right = getSubtree(node->right, curDepth + 1, targetDepth);
        
        if (left && right) return node;
        if (left) return left;
        return right;
    }
    
    TreeNode* subtreeWithAllDeepest(TreeNode* root) {
     
        if (!root->left && !root->right) return root;
        return getSubtree(root, 0, findDepth(root));
    }
};

File: 865-smallest-subtree-with-all-the-deepest-nodes/865-smallest-subtree-with-all-the-deepest-nodes.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findDepth(self, node: TreeNode) -> int:
        
        if not (node.left or node.right): return 0
        
        ldepth =  self.findDepth(node.left) if node.left else 0
        rdepth = self.findDepth(node.right) if node.right else 0
        return 1 + max(ldepth, rdepth)
    
    def findSubtree(self, node: TreeNode, curDepth: int, targetDepth: int) -> TreeNode:
        
        if curDepth == targetDepth or not node: return node
        #if not (node.left or node.right): return None
        
        lnode = self.findSubtree(node.left, curDepth + 1, targetDepth)
        rnode = self.findSubtree(node.right, curDepth + 1, targetDepth)
        
        if lnode and rnode: return node
        elif lnode: return lnode
        else: return rnode
        
    def subtreeWithAllDeepest(self, root: TreeNode) -> TreeNode:
        
        depth = self.findDepth(root)
        
        if not depth: return root

        return self.findSubtree(root, 0, depth)
            
            